nemo/collections/asr/models/enhancement_models.py

```diff
@@ -464,3 +199,380 @@ def list_available_models(cls) -> Optional[PretrainedModelInfo]:
+            "input_signal": NeuralType(('B', 'C', 'T'), AudioSignal()),
+            "input_length": NeuralType(tuple('B'), LengthsType()),
+        },
+        output_types={"loss": NeuralType(None, LossType()),},
+    )
+    def _step(self, target_signal, input_signal, input_length=None):
+        """Randomly generate a time step for each example in the batch, estimate
+        the score and calculate the loss value.
+
+        Note that this step does not include sampler.
+        """
+        batch_size = target_signal.size(0)
+
+        if self.normalize_input:
+            # max for each example in the batch
+            norm_scale = torch.amax(input_signal.abs(), dim=(-1, -2), keepdim=True)
+            # scale input signal
+            input_signal = input_signal / (norm_scale + self.eps)
+            # scale the target signal
+            target_signal = target_signal / (norm_scale + self.eps)
+
+        # Apply encoder to both target and the input
+        input_enc, input_enc_len = self.encoder(input=input_signal, input_length=input_length)
+        target_enc, _ = self.encoder(input=target_signal, input_length=input_length)
+
+        # Generate random time steps
+        sde_time = self.sde.generate_time(size=batch_size, device=input_enc.device)
+
+        # Get the mean and the variance of the perturbation kernel
+        pk_mean, pk_std = self.sde.perturb_kernel_params(state=target_enc, prior_mean=input_enc, time=sde_time)
+
+        # Generate a random sample from a standard normal distribution
+        z_norm = torch.randn_like(input_enc)
+
+        # Prepare perturbed data
+        perturbed_enc = pk_mean + pk_std * z_norm
+
+        # Score is conditioned on the perturbed data and the input
+        estimator_input = torch.cat([perturbed_enc, input_enc], dim=-3)
+
+        # Estimate the score using the neural estimator
+        # SDE time is used to inform the estimator about the current time step
+        # Note:
+        # - some implementations use `score = -self._raw_dnn_output(x, t, y)`
+        # - this seems to be unimportant, and is an artifact of transfering code from the original Song's repo
+        score_est, score_len = self.estimator(input=estimator_input, input_length=input_enc_len, condition=sde_time)
+
+        # Score loss weighting as in Section 4.2 in http://arxiv.org/abs/1907.05600
+        score_est = score_est * pk_std
+        score_ref = -z_norm
+
+        # Score matching loss on the normalized scores
+        loss = self.loss(estimate=score_est, target=score_ref, input_length=score_len)
+
+        return loss
+
+    # PTL-specific methods
+    def training_step(self, batch, batch_idx):
+        input_signal, input_length, target_signal, target_length = batch
+
+        # For consistency, the model uses multi-channel format, even if the channel dimension is 1
+        if input_signal.ndim == 2:
+            input_signal = input_signal.unsqueeze(1)
```

Comment:
Nitpick: would be really nice to have this as `einops.rearrange` call, this would document dimensions.

Same for other `squeeze`/`unsqueeze` calls inside methods like `training_step`